[4.0] Debug bar

elkuku opened the pull request:
### Summary of Changes

The goal of this PR is to refactor the debug plugin (currently more than 2.000 lines of code) into smaller pieces.
An external library is used to display a floating debug bar. [See a demo](http://phpdebugbar.com/)

### Testing Instructions

<del>run `composer install` after switching to this branch
I have not included any vendor files or modified composer files here to make reviewing easier. Is that correct?</del>
Turn on debug and/or language debug and play with the various options in plugin configuration.
Do **not** turn on deprecation logging since the logger is using currently deprecated API creating an infinite loop...

### Expected result

A beautiful debug toolbar :wink: 

### Actual result

A not so beautiful looong thingy, fixed to the very bottom of the page.

### Documentation Changes Required
Alot :tongue: 

### Extended Description

While working on a [Symfony](https://symfony.com) project, I was amazed by the debug bar they offer and I thought about the old Joomla! plugin...<br />
Integrating the Symfony bar would require a lot of dependencies from Symfony, probably including the Symfony Application class. While this would be amazing, I suspect that this is not gonna happen any time soon...

So I started playing with [maximebf/debugbar](https://github.com/maximebf/php-debugbar) which claims to be inspired by the Symfony debug bar, but for "non Symfony" projects `;)`

Thise resulted in some `DataCollector` classes which, apart from collecting data, are responsible for bringing the collected data to something "serializable".

This data is then serialized and written to disk for later analysis.

After a thousand words, here are some screens:
**Note**: Since the "old" code is still in place, you will see the "old" output as well as the "new " output in the screenshots.

### Session
Currently the session display in debug plugin is broken.
![2017-04-18-121246_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298253/2494a056-26b9-11e7-9665-c3a25c06b794.png)

### Profile
![2017-04-21-133031_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298255/249a030c-26b9-11e7-8b95-0a6abbf31da6.png)

### Queries
Only the database queries are displayed. Extended display should happen on a dedicated page.
![2017-04-18-121436_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298254/2496ae32-26b9-11e7-81e0-8ff85bf51d2b.png)

### Deprecation logging
Seems currently broken since file loggers are already using deprecated API, creating funny infinite loops...
![2017-04-21-171552_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298260/24d1f938-26b9-11e7-90b1-7e585fa01473.png)


### Language stuff
Should look as before, just more readable and compact.

#### Loaded
![2017-04-21-132535_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298256/249d4558-26b9-11e7-90a3-1d8f15842a81.png)

#### Errors 
![2017-04-18-121557_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298257/249d6ae2-26b9-11e7-8406-8ef19db5b84b.png)

#### Untranslated
![2017-04-18-121539_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298258/249e1d2a-26b9-11e7-87c6-8a563fd82965.png)

### Hidden errors
While playing here I discovered a query error that is so much eaten up and swallowed that it does not appear in any way on the standard output. I would bet that this is causing a bug...
![2017-04-21-171402_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/25298259/24ca8b62-26b9-11e7-8588-da31d01b31ad.png)

### Mobile ready
![screenshot_20170426-010917](https://cloud.githubusercontent.com/assets/33978/25421343/054e2354-2a22-11e7-8501-256bd4e49acc.png)

Tested on a 5'' screen with fat fingers and it worked pretty well.

----

Happy debugging `=;)`

elkuku committed:
Add data collectors

elkuku committed:
Update debug plugin

elkuku committed:
Update composer.json
Add maximebf/debugbar

elkuku committed:
Fix CS

elkuku committed:
Merge branch '4.0-dev' into debug-bar

elkuku committed:
Add log messages

elkuku committed:
Add a data formatter

elkuku committed:
Add more data collectors

elkuku committed:
Use the data formatter

elkuku committed:
Add custom widgets

elkuku committed:
Cleanup

elkuku committed:
Remove obsolete method

zero-24 commented:
> run composer install after switching to this branch
I have not included any vendor files or modified composer files here to make reviewing easier. Is that correct?

@elkuku please include the composer files in your commit and place them in the vendor folder. As we are shipping the vendor files with the CMS ;)

elkuku commented:
@zero-24 that would be over 6.000 files, making it almost impossible for reviewing. Please reconsider..

BTW...
> place them in the vendor folder.

I believe you mean `/libraries/vendor`?

zero-24 commented:
Yes that folder. Hmm can you .gitignore files out that are not needed?

As without shipping than with the cms we cant use that lib in the current architecture.

mbabker commented:
Run "Composer install", "Composer update `package`", "Composer install --no-dev" and you'll be good.

elkuku committed:
Add vendor files
Update composer files

elkuku commented:
@zero-24 I understand that vendor files have to be included in the final package, I just thought it would be easier to review...

@mbabker Of course your're right (as usual), most of the files are `dev` dependencies...

Anyway, I created a new branch, you may review the diff [here](https://github.com/joomla/joomla-cms/compare/4.0-dev...elkuku:debug-bar-with-vendors?expand=1), and if you find it reviewable, I'll merge it in here...

Bakual commented:
The needed dependencies should be "reviewed" as well anyway. Not exactly codewise but personally I want to see what you're going to pull in.
For example If it needs 5000 lines of code in various dependencies to replace the 2000 lines in our debug plugin, then we at least should question it. 😊

brianteeman commented:
Files not lines

Bakual commented:
The 2000 files are with dev stuff which we don't need. So the actual file count is lower. I just did a guess with the 5000 lines. Looking at the other branch, it's actually 20'000 lines we add. So imho there has to be a very good reason to add all those dependencies just for a debug plugin.
But that is my personal opinion.

wilsonge commented:
Can we use the namespacing convention at https://volunteers.joomla.org/teams/joomla-4-architecture/reports/115-joomla-4-architecture-sprint-odense-dk please. So `Joomla\Plugin\System\Debug` please :)

wilsonge commented:
I like this concept overall :)

elkuku committed:
Use the temp folder to store request data
Introduce com_profiler =)

elkuku committed:
Remove obsolete method - since JError is @deprecated 1.7

elkuku committed:
Remove obsolete methods - Session, Memory and Language stuff

wilsonge commented on plugins/system/debug/debug.php:
```diff
@@ -120,6 +120,17 @@ public function __construct(&$subject, $config)
+			$this->app = JFactory::getApplication();
+		}
+
+		if ('com_profiler' == $this->app->input->get('option'))
```
Why do we have something for community builder here?

elkuku replied:
LOL I could have sworn that I heard com_profiler before... sorry I almost left the Joomla! community :(

And no, it's not community builder I have in mind here but a component that will display the extended "profile" information that the debug bar is unable to display (due to its size).

It has slipped in here and will be removed.

elkuku commented:
@Bakual and @brianteeman 
> The needed dependencies should be "reviewed" as well anyway. Not exactly codewise but personally I want to see what you're going to pull in.

I thought the change in composer.json might be enough, but you are right. So here are some more infos on files and lines :wink: 

### php-debugbar
* https://github.com/maximebf/php-debugbar
* http://phpdebugbar.com/

```
~/r/joomla-cms (debug-bar) $ phploc libraries/vendor/maximebf/
phploc 3.0.1 by Sebastian Bergmann.

Directories                                          7
Files                                               45

Size
  Lines of Code (LOC)                             6328
  Comment Lines of Code (CLOC)                    2123 (33.55%)
  Non-Comment Lines of Code (NCLOC)               4205 (66.45%)
  Logical Lines of Code (LLOC)                    1179 (18.63%)
```
The library has one dependency:

### symfony vardumper
* https://github.com/symfony/var-dumper
* http://symfony.com/doc/current/components/var_dumper.html

```
~/r/joomla-cms (debug-bar) $ phploc libraries/vendor/symfony/var-dumper/
phploc 3.0.1 by Sebastian Bergmann.

Directories                                         10
Files                                               56

Size
  Lines of Code (LOC)                             7620
  Comment Lines of Code (CLOC)                    1074 (14.09%)
  Non-Comment Lines of Code (NCLOC)               6546 (85.91%)
  Logical Lines of Code (LLOC)                    1708 (22.41%)
```
This is a pretty cool replacement for `var_dump()` and can be used also by 3PDs.

elkuku commented:
@wilsonge 
> I like this concept overall :)

Big Thanks for this :wink: 

Now on the namespacing thing.. I'd love to follow any conventions but I wasn't aware of the existence..
Could you give me some advice on how to accomplish this?
I mean I would have to create the folder(s) `JROOT/plugins/system/debug/Joomla/Plugin/System/Debug` and register the plugin root as the namespace root?
That seems odd :(

elkuku committed:
Remove test code

wilsonge commented:
No there is going to be a PSR-4 map of Joomla/Plugin/System/Debug to JROOT/plugins/system/debug :) (see https://github.com/joomla/joomla-cms/pull/15226)

elkuku committed:
Review plugin parameters
Remove additional language debug options
Remove db log option since everything is logged by default =;)

Bakual commented:
@elkuku Please just do the PR with all the code that will be included with your PR. There is no point merging a PR when we have to manually run composer afterwards to get it working. It has to work right away. It's both easier to test and maintain and less failure will happen.

zero-24 commented:
Expecial as currently travis is failing because of the missing code ;) 

```
.....S...PHP Fatal error:  Class 'DebugBar\DebugBar' not found in /home/travis/build/joomla/joomla-cms/plugins/system/debug/debug.php on line 202
Fatal error: Class 'DebugBar\DebugBar' not found in /home/travis/build/joomla/joomla-cms/plugins/system/debug/debug.php on line 202
```

https://travis-ci.org/joomla/joomla-cms/jobs/224803602

elkuku commented:
All vendor files are commited, unit- and system tests run fine both locally and online and all CI tools are "happy" :wink: 
I wonder why the debug plugin is even executed in CI test but there may be reasons...

@wilsonge I am following the other PR and hope that it will be merged in a timely manner... But until than how would I setup the namespace outlined in the doc you mentioned?
I would be happy to change it to anything that works, since the namepace I picked here is an ugly band aid until you guys decide about a "proper" way :tongue:

elkuku committed:
Use a standard namespace - Thanks @wilsonge

elkuku commented:
@wilsonge OK I found it finally ;)

elkuku commented:
Hmmm Mrs. joomla-cms-bot changed the labels here on JIssues...
`External Library  JavaScript  Front End  Plugins	⇒	External Library  Libraries`

Call the devs!! :tongue: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15472">issues.joomla.org/tracker/joomla-cms/15472</a>.</sub>

wilsonge commented:
I think this is going to break once I merge 3.7 stable into 4.0. One of the changes has been to via htaccess block direct access to libraries/vendor (exactly like Drupal do https://github.com/drupal/drupal/blob/8.4.x/core/lib/Drupal/Core/Composer/Composer.php#L106-L137). However this is going to have the effect here of blocking access to the js and css files required for this to work

elkuku committed:
Add missing icons

elkuku commented:
Cool. I would have actually preferred moving those library files out of the web root completely but yeah...

OK, so I added a line to `Gruntfile.js` to copy all the relevant files (more or less) to `media/vendor`.
Now I suppose that I should also commit those files? Oo 

In the meantime I commit the Grutfile. @wilsonge Could you check if the changes are appropriate, please?

elkuku committed:
Update Gruntfile.js
Add copy Debug Bar files

elkuku committed:
Split deprecation logging
Review parameters

elkuku committed:
Use the "new" classes

elkuku commented:
Not sure why the drone test is failing with
```
27 04 2017 05:28:09.226:ERROR [Firefox 50.0.0 (Ubuntu 0.0.0)
 | Validate, validate method on #validate-numeric-nan | encountered a declaration exception]: 
TypeError: Joomla.optionsStorage is null in 
http://localhost:9876/base/media/system/js/core.js?0fd424e7fb8213be9940ef87c205dba3db7df6d0 (line 9)
```
THB I have no idea why this might fail and I suspect that it's not related to the changes I made.
Advise please.

dgt41 commented:
@elkuku it's not your code, it's validator that blindly asks for a string and it throws if it's not declared. I'll fix that later on...

elkuku commented:
@dgt41 Thanks, good to know :wink:

elkuku committed:
The 'Cms' is a 'CMS' now...

elkuku committed:
Remove a local variable

elkuku committed:
Rename data collector classes

elkuku committed:
Add vendor files in /media/vendor

Fedik commented:
guys what need to do to get this accepted? 
it seems a great improvement

dgrammatiko commented:
Personally I would go with: https://underground.works/clockwork/
https://github.com/itsgoingd/clockwork

Fedik commented:
> I would go with: https://underground.works/clockwork/

only drawback, it forces me to install the browser extension 😄

brianteeman commented:
> or use the web UI http://your.app/__clockwork

mbabker commented:
They need to add a LICENSE file to that repo.  The note of the license in the Composer manifest isn't enough.

Fedik commented:
> or use the web UI http://your.app/__clockwork

did not seen this part :)

brianteeman commented:
@mbabker I wondered about that

brianteeman commented:
It does state MIT in the composer file

Fedik commented:
sooo, I have tested both [Debug bar](https://github.com/maximebf/php-debugbar) and [Clockwork](https://github.com/itsgoingd/clockwork)

In 2 words, for Joomla! I would choose "Debug bar".


Here is some notes/comparison:

**Debug bar**

**pros:**
- client side rendering (based on own renderer)
- possible to view debug history (not implemented in current pull)
- possible to create a custom data handlers
- possible to create a custom client side widgets
- not bad documentation

**cons:**
- I didn't found anything critical


**Clockwork**

**pros:**
- client side rendering (based on AgularJS 1.5)
- possible to view debug history
- possible to create a custom data handlers


**cons:**
- client side depend from browser extension or need to adapt /Clockwork/Web/public
- not possible to create a custom client side widgets, except we write custom client rendering
- documentation does not exist (I not found)


So I think **Clockwork** is a nice toy or when someone have a time to fork and adapt it,
for Joomla! I would choose **Debug bar**.

Fedik commented:
@elkuku I have fixed the conflicts https://github.com/Fedik/joomla-cms/tree/debug-bar  

Now I think, it would be more easy to write new plugin than rewrite existing 😄

laoneo commented:
I would welcome to have a more modern debug bar for Joomla. Can you bring that pr into a state for reevaluation please. Thanks!

wilsonge commented:
There's no action happening here. i'd love to see something like this make it into 4.0 - but I guess it's not happening in this PR - so I'm going to close this for now. Of course if it's comes back in sync we can reopen and evaluate :)

elkuku committed:
merging

elkuku committed:
Update vendor files

elkuku committed:
Update vendor files

elkuku committed:
Update debug plugin

elkuku commented:
This has been updated

elkuku commented:
Sorry but it seems that something got wrong with the last merge... I'd like to close and redo this PR but unfortunately I am unable to close it....
